tty: use blocking mode on OS X

Fishrock123 opened the pull request:
<!--
Thank you for your pull request. Please review below requirements and walk
through the checklist. You can 'tick' a box by using the letter "x": [x].

Run the test suite with: `make -j4 test` on UNIX or `vcbuild test nosign` on
Windows.

If this aims to fix a regression or you’re adding a feature, make sure you also
write a test. If possible, include a benchmark that quantifies your changes.

Finally, read through our contributors guide and make adjustments as necessary:
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- remove lines that do not apply to you -->

- [X] tests and code linting passes
- [x] a test and/or benchmark is included
- [ ] documentation is changed or added
- [X] the commit message follows commit guidelines


##### Affected core subsystem(s)
<!-- provide affected core subsystem(s) (like doc, cluster, crypto, etc) -->
tty, process (OS X)

##### Description of change
<!-- provide a description of the change below this comment -->

Refs: https://github.com/nodejs/node/pull/1771
Refs: https://github.com/nodejs/node/issues/6456
Refs: https://github.com/nodejs/node/pull/6773
Refs: https://github.com/nodejs/node/pull/6816

This is a quicker fix on the most broken platform, OS X, while we investigate what needs to be done on other platforms.

The goal is to get **at least** this into next week's v6 release.

I will try to pull together a test later today. This restores the original, (now) expected, OS X behavior.

Fishrock123 commented:
@nodejs/ctc I'm resorting to this in case we are not able to determine what to do across all platforms for next week's release.

CI: https://ci.nodejs.org/job/node-test-pull-request/2712/

jasnell commented:
I'm ok with this as a short term solution.

Fishrock123 commented:
If anyone could help me with some python for the tests that would be great! Testing branch at https://github.com/Fishrock123/node/tree/test-tty-issues

cc @nodejs/build maybe? @thefourtheye? also cc @addaleax who pointed me to `pty.openpty()`.

Fishrock123 commented:
Updated with the WIP testcases. The special test runner config still stalls out for me though. :/

thefourtheye commented on test/pseudo-tty/testcfg.py:
```diff
@@ -0,0 +1,159 @@
+    else: return str.startswith('==') or str.startswith('**')
+
+  def IsFailureOutput(self, output):
+    f = file(self.expected)
```
Well, preferred way is to use `open`.

Fishrock123 replied:
copied verbatim from `test/message/testcfg.py`

thefourtheye commented on test/pseudo-tty/testcfg.py:
```diff
@@ -0,0 +1,159 @@
+      print "expect=%d" % len(patterns)
+      print "actual=%d" % len(outlines)
+      print "patterns:"
+      for i in xrange(len(patterns)):
```
Iterate `patterns` and `outlines` directly.

For example,

```python
for pat in patterns:
  ...
```

Fishrock123 replied:
also copied verbatim from `test/message/testcfg.py`

thefourtheye commented on test/pseudo-tty/testcfg.py:
```diff
@@ -0,0 +1,159 @@
+        print "outline = %s" % outlines[i]
+      return True
+    for i in xrange(len(patterns)):
+      if not re.match(patterns[i], outlines[i].strip()):
```
stripping can be done while creating outlines itself

thefourtheye commented on test/pseudo-tty/testcfg.py:
```diff
@@ -0,0 +1,159 @@
+      print "outlines:"
+      for i in xrange(len(outlines)):
+        print "outline = %s" % outlines[i]
+      return True
```
Shouldn't this be False?

Fishrock123 commented:
@thefourtheye Any idea why it doesn't work? :S

Fishrock123 commented:
Also cc @isaacs to see if he is ok with this solution for now?

thefourtheye commented:
@Fishrock123 I am working on fixing the stalling test... I forgot to update here... Sorry :D

thefourtheye commented:
@Fishrock123 Can you try this patch.

```diff
diff --git a/tools/test.py b/tools/test.py
index 53909ec..874e239 100755
--- a/tools/test.py
+++ b/tools/test.py
@@ -576,11 +576,17 @@ def RunProcess(context, timeout, args, **rest):
       error_mode = SEM_NOGPFAULTERRORBOX;
       prev_error_mode = Win32SetErrorMode(error_mode);
       Win32SetErrorMode(error_mode | prev_error_mode);
+
+  faketty = rest.pop('faketty', False)
+  pty_out = rest.pop('pty_out')
+
   process = subprocess.Popen(
     shell = utils.IsWindows(),
     args = popen_args,
     **rest
   )
+  if faketty:
+    os.close(rest['stdout'])
   if utils.IsWindows() and context.suppress_dialogs and prev_error_mode != SEM_INVALID_VALUE:
     Win32SetErrorMode(prev_error_mode)
   # Compute the end time - if the process crosses this limit we
@@ -592,6 +598,31 @@ def RunProcess(context, timeout, args, **rest):
   # loop and keep track of whether or not it times out.
   exit_code = None
   sleep_time = INITIAL_SLEEP_TIME
+  if faketty:
+    output = ''
+    while True:
+      if time.time() >= end_time:
+        # Kill the process and wait for it to exit.
+        KillProcessWithID(process.pid)
+        exit_code = process.wait()
+        timed_out = True
+        break
+
+      # source: http://stackoverflow.com/a/12471855/1903116
+      # related: http://stackoverflow.com/q/11165521/1903116
+      try:
+        data = os.read(pty_out, 9999)
+      except OSError as e:
+        if e.errno != errno.EIO:
+          raise
+        break # EIO means EOF on some systems
+      else:
+        if not data: # EOF
+          break
+        output += data
+
+    return process, process.poll(), timed_out, output
+
   while exit_code is None:
     if (not end_time is None) and (time.time() >= end_time):
       # Kill the process and wait for it to exit.
@@ -604,7 +635,7 @@ def RunProcess(context, timeout, args, **rest):
       sleep_time = sleep_time * SLEEP_TIME_FACTOR
       if sleep_time > MAX_SLEEP_TIME:
         sleep_time = MAX_SLEEP_TIME
-  return (process, exit_code, timed_out)
+  return (process, exit_code, timed_out, '')


 def PrintError(str):
@@ -630,27 +661,30 @@ def Execute(args, context, timeout=None, env={}, faketty=False):
   if faketty:
     (out_master, fd_out) = pty.openpty()
     fd_err = fd_out
+    pty_out = out_master
   else:
     (fd_out, outname) = tempfile.mkstemp()
     (fd_err, errname) = tempfile.mkstemp()
+    pty_out = None

   # Extend environment
   env_copy = os.environ.copy()
   for key, value in env.iteritems():
     env_copy[key] = value

-  (process, exit_code, timed_out) = RunProcess(
+  (process, exit_code, timed_out, output) = RunProcess(
     context,
     timeout,
     args = args,
     stdout = fd_out,
     stderr = fd_err,
-    env = env_copy
+    env = env_copy,
+    faketty = faketty,
+    pty_out = pty_out
   )
+
   if faketty:
-    output = os.read(out_master, 9999)
     os.close(out_master)
-    os.close(fd_out)
     errors = ''
   else:
     os.close(fd_out)
```

Fishrock123 commented:
Looks like github is still having problems. I'll run a Ci when the branch updates properly. Thanks for all the help @thefourtheye, it passes for me locally, and fails when the patch is removed. :)

Fishrock123 commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/2745/

Fishrock123 commented:
Crap, looks like `pty` can't be imported on windows... https://ci.nodejs.org/job/node-test-binary-windows/2235/

@thefourtheye Any ideas?

thefourtheye commented:
@Fishrock123 Python 2.x standard library itself doesn't support `pty` in Windows. I was working with Antti Haapala on this. So cc @ztane.

Can we just skip these tests in Windows?

Fishrock123 commented:
> Can we just skip these tests in Windows?

I would say yes, but given the functionality needs to be part of the test runner I'm not sure how?

thefourtheye commented:
We can either do this

```python
diff --git a/test/pseudo-tty/testcfg.py b/test/pseudo-tty/testcfg.py
index 1fdc35a..2e21902 100644
--- a/test/pseudo-tty/testcfg.py
+++ b/test/pseudo-tty/testcfg.py
@@ -29,6 +29,8 @@ import test
 import os
 from os.path import join, dirname, exists, basename, isdir
 import re
+import utils
+

 FLAGS_PATTERN = re.compile(r"//\s+Flags:(.*)")

@@ -134,6 +136,9 @@ class TTYTestConfiguration(test.TestConfiguration):
   def ListTests(self, current_path, path, arch, mode):
     all_tests = [current_path + [t] for t in self.Ls(self.root)]
     result = []
+    # Skip these tests on Windows, as pseudo terminals are not available
+    if utils.IsWindows():
+      print "Skipping pseudo-tty tests, as pseudo terminals are not available in Windows
+      return result
     for test in all_tests:
       if self.Contains(path, test):
         file_prefix = join(self.root, reduce(join, test[1:], ""))
```

or this

```python
diff --git a/tools/test.py b/tools/test.py
index 28fc39b..71e68fb 100755
--- a/tools/test.py
+++ b/tools/test.py
@@ -1519,13 +1519,8 @@ def Main():
   repositories += [TestRepository(a) for a in options.suite]

   root = LiteralTestSuite(repositories)
-  if len(args) == 0:
-    paths = [SplitPath(t) for t in BUILT_IN_TESTS]
-  else:
-    paths = [ ]
-    for arg in args:
-      path = SplitPath(arg)
-      paths.append(path)
+  paths = [SplitPath(t) for t in args or BUILT_IN_TESTS
+           if not(t == 'psuedo-tty' and utils.IsWindows())]

   # Check for --valgrind option. If enabled, we overwrite the special
   # command flag with a command that uses the run-valgrind.py script.
```

Fishrock123 commented:
@thefourtheye It's failing on **import** though:

```
Traceback (most recent call last):
  File "tools\test.py", line 46, in <module>
    import pty
```

thefourtheye commented:
Ah, that can fixed by simply moving the `import pty` to the line above `... = pty.openpty()`

thefourtheye commented on test/pseudo-tty/testcfg.py:
```diff
@@ -0,0 +1,159 @@
+
+import test
+import os
+from os.path import join, dirname, exists, basename, isdir
```
`dirname` is not used.

Fishrock123 commented:
Updated with suggestions, new CI: https://ci.nodejs.org/job/node-test-pull-request/2749/

Fishrock123 commented:
Updated, added the tests to `test-ci`. Another run: https://ci.nodejs.org/job/node-test-pull-request/2751/

kzc commented:
So Mac and Windows will have blocking ttys at start up, but Linux and other unixes will not? That would only encourage non-portable application behavior.

What is the downside to simply making all platforms have blocking writes to ttys at start up?

Fishrock123 commented:
> So Mac and Windows will have blocking ttys at start up, but Linux and other unixes will not? 

I've gotten mixed messages as to whether Windows is actually blocking to a console, so I'm not actually sure.

> That would only encourage non-portable application behavior

That's already the case. At this point we are only trying to fix broken things.

> What is the downside to simply making all platforms have blocking writes to ttys at start up?

Some people are concerned about linux.

kzc commented:
> At this point we are only trying to fix broken things.

"Broken" implies the 5.x behavior is correct. It is not. 5.x behaves quite differently on Mac and Linux. Linux only outputs up to 64KB to the console upon `process.exit()`, while 5.x on Mac does not experience this truncation problem due to blocking tty writes as we now know.

> Some people are concerned about linux.

As am I. node 5.x on Linux has a number of other truncation issues pertaining to ttys well. For example, exceptions are truncated with node 5.x on Linux:

https://github.com/nodejs/node/issues/6867#issuecomment-220174747

Rather than restoring unintended legacy behaviors on certain platforms node should take this opportunity to unify the behavior on all platforms.

Fishrock123 commented:
> "Broken" implies the 5.x behavior is correct.

Correct or not, the v6 behavior causes far more problems.

> As am I. node 5.x on Linux has a number of other truncation issues pertaining to ttys well. For example, exceptions are truncated with node 5.x on Linux:

Some people are concerned about blocking on linux for performance or other reasons which **I** am unable to investigate all at once.

> Rather than restoring unintended legacy behaviors on certain platforms node should take this opportunity to unify the behavior on all platforms.

I don't disagree but I would rather fix the majority of broken problems asap.

Fishrock123 commented:
Hmmm, i don't think the test runner is working properly on linux/smartos/freebsd...

The following is the correct output (as far as I can tell), but it's seeing it as a failure?
https://ci.nodejs.org/job/node-test-commit-smartos/2608/nodes=smartos14-64/console
```
not ok 1119 pseudo-tty/no_dropped_stdio
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# ooooooooooooooooooooooooooooooooooooO
  ---
  duration_ms: 0.276
  ...
not ok 1120 pseudo-tty/no_interleaved_stdio
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# oooooooooooooooooooooooooooooooooooooooooooooooooo
# ooO
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
# xxX
  ---
  duration_ms: 0.284
  ...
```

Fishrock123 commented:
cc @bnoordhuis, do you know much about the message test runner works? You seem to be the last person with a commit to it..

Fishrock123 commented:
Can't seem to repo on my ubuntu machine.. `Linux ubuntu-tower 4.2.0-25-generic #30-Ubuntu SMP Mon Jan 18 12:31:50 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux`

Fishrock123 commented:
Another CI run... https://ci.nodejs.org/job/node-test-pull-request/2766/

Will maybe try upgrading my ubuntu to 16.04 later today...

Fishrock123 commented:
Canceled that run. Going to force it to produce error output and see what happens.. https://ci.nodejs.org/job/node-test-pull-request/2767/

bnoordhuis commented:
> cc @bnoordhuis, do you know much about the message test runner works?

If you have specific questions, I'll try to answer them.

Fishrock123 commented:
Stalled because I have no idea what is going on with the test runner.

I think @TheAlphaNerd took a bit of a look yesterday and found that python sub-processes were crashing or something? I'm really not sure.

rvagg commented:
builds are available here to try out https://nodejs.org/download/rc/v6.2.1-rc.1/

rvagg commented on test/pseudo-tty/no_dropped_stdio.js:
```diff
@@ -0,0 +1,13 @@
+
+var out = new Array(bytes).join('o');
+out += 'O'; // End ID character
+out = out.replace(/(o{50})/g, '$1\n');
```
This results in 1007 bytes which doesn't really do the trick. Can I suggest changing the above 5 lines to:

```
// create 1052 bytes of output, wrapped at 50 columns, terminated with an 'O'
// enough to overflow the 1kb OS X TTY buffer
var out = new Array(22).join(new Array(50).join('o') + '\n') + 'O\n'
```

Then the no_dropped_stdio.out would need to become:

```
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
O
```

rvagg commented on test/pseudo-tty/no_interleaved_stdio.js:
```diff
@@ -0,0 +1,18 @@
+
+var err = new Array(bytes).join('x');
+err += 'X'; // End ID character
+err = err.replace(/(x{50})/g, '$1\n');
```
likewise:

```js
// create 1052 bytes of output, wrapped at 50 columns, terminated with an 'O'
// enough to overflow the 1kb OS X TTY buffer
var out = new Array(22).join(new Array(50).join('o') + '\n') + 'O\n'
var err = new Array(22).join(new Array(50).join('x') + '\n') + 'X\n'
```

and no_interleaved_stdio.out would be:

```
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
ooooooooooooooooooooooooooooooooooooooooooooooooo
O

xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
X
```

rvagg commented:
the code change lgtm, sans comments about output length but I know you're working through that will come to an appropriate solution there. The outstanding things are the failures on other platforms.

I ran another CI just to get some clarity: https://ci.nodejs.org/job/node-test-commit/3530/

You can't see it, the output from most of the failures I _think_ is `--- CRASHED ---` for both of the tests, stderr I guess, I had to run manually on freebsd and ubuntu14.04 slaves to see that output and I suppose it's to do with fake tty stuff and I can't reproduce locally even though I'm testing this on Ubuntu 16.04 x64. I guess you're having the same problem! I thought maybe it was to do with the slow connection to the machines interacting with remote tty somehow but I can't replicate that either when I route an ssh session back to my local machine through a remote one.

Fishrock123 commented:
I've changed the test cases to be even more exact. I think they output exactly the correct buffer sizes this time. (1025 bytes).

rvagg commented:
I meant to add extra details to my lgtm:
* I tested and confirmed this locally on Linux and OSX, running the tests before and after the code change and it all works as expected. Borked CI is the missing piece. 
* I approve of this change in behaviour even though it's technically semver-major because it's fixing a change in the way stdio works on OSX between 5 and 6 that was unintentionally induced by a libuv change. We've been fudging the intended behaviour till now and this is just returning us to that state.
* I suspect this is not a permanent fix, it's certainly not a universal one because it's only addressing OSX and Linux still has different behaviour but the buffer size makes the user experience different. We can work on something more comprehensive for the whole stdio mess in #6980 and that may involve revisiting and adjusting this change but in comfortable moving forward with this as is for now.

Fishrock123 commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/2823/

kzc commented:
> it's fixing a change in the way stdio works on OSX between 5 and 6 that was unintentionally induced by a libuv change

Actually OSX stdout tty was historically blocking from node 0.10 to node 5.x by accident in libuv. libuv 1.9.0 fixed an unrelated bug related to redirecting ttys which made Mac stdio tty fds non-blocking in node 6.0.0 to match the behavior of all other unix platforms. This non-blocking behavior exposed the stdout truncation issue on Mac, and the lack of flushing of stdout and stderr upon process.exit()/uncaught exceptions.

Fishrock123 commented:
@kzc you and @rvagg are talking about the same thing.

Fishrock123 commented:
Updated to only run the pty tests on OS X, new ci... https://ci.nodejs.org/job/node-test-pull-request/2825/

This isn't ideal, since these tests ought to pass on all non-windows platforms, but this is really as much as I'm capable of in terms of python. :(

Fishrock123 commented:
Still failing... but the tests shouldn't even run now? I'm confused.

Fishrock123 commented on test/pseudo-tty/testcfg.py:
```diff
@@ -0,0 +1,166 @@
+    # Skip these tests on Windows, as pseudo terminals are not available
+    # FIXME(Fishrock123): Currently this only runs properly on OS X in the
+    # Jenkins CI
+    if platform.system() is 'Darwin':
```
@thefourtheye / @jbergstroem Is this not correct?

addaleax replied:
Forgive me if I’m wrong, but shouldn’t this be `is not 'Darwin'`?

Fishrock123 replied:
_oh_

thefourtheye replied:
In Python, we should never use `is` operator to compare strings.

Fishrock123 commented:
Ok it should now properly only run on OS X... https://ci.nodejs.org/job/node-test-pull-request/2827/

saghul commented on lib/tty.js:
```diff
@@ -48,6 +48,13 @@ function WriteStream(fd) {
     writable: true
   });
 
+  // Prevents interleaved stdout/stderr output in OS X terminals.
```
I don't think this comment is correct. The change is intended to fix loosing output when doing process.exit. Interleaving will depend on OS level buffering. Here is the full backstory in case you want to link to it: https://github.com/nodejs/node/issues/6806#issuecomment-222350923

saghul replied:
To clarify, IMHO the interleaving is a side effect of the blocking-ness.

Fishrock123 replied:
> The change is intended to fix loosing output when doing process.exit.

Technically it is intended to fix __both__.

>  Interleaving will depend on OS level buffering.

Of course, and OS X is hardcoded to 1kb so it won't change regardless.

> To clarify, IMHO the interleaving is a side effect of the blocking-ness.

Sure, but it's still a problem we need to cover.

Typically, it would just be an annoyance. You'd have all the information, just sorta mixed up.

But that's not always the case, since people often use ansi control characters in CLI apps.

Consider a problem I ran into while writing the test: At 1024kb of data, including the newline `\n`, an entire line of data no longer appears.

Why?

Somewhere down the line `\n` is being changed to `\n\r`, which overflows the buffer by one character, leaving a separate `\r` which then erases the current line. Who is going to expect that?

If not me, very few people indeed. (That was done using `Terminal.app` on OS X. Given it is the default terminal, this will be in no way isolated.)

saghul replied:
Ah, I see. Carry on then! :-)

kzc commented:
>> The change is intended to fix loosing output when doing process.exit.

> Technically it is intended to fix **both**.

Not the case. Only fixes OSX tty output only upon process.exit(). Doesn't fix other unix platforms when OS stdio buffer exceeded or when stdio is piped.

Fishrock123 commented:
@kzc Please read the context. The code comment says `... in OS X terminals.`

addaleax commented:
@Fishrock123 addaleax/node@27c3ca99354dd85c93c296ca24b26a5430b46499 should fix the Linux flakyness (and makes sense in general, I think), if you’re interested

Fishrock123 commented:
will test when jenkins is back online I suppose

addaleax commented:
it seems to be, CI: https://ci.nodejs.org/job/node-test-commit/3576/

Fishrock123 commented:
Hmm, looks like there were some cancellations in there. Running again: https://ci.nodejs.org/job/node-test-pull-request/2863/

Fishrock123 commented:
Hmm, looks like ARM is still exploding due to hardware issues. Tests for this seems to pass fine.

@addaleax Mind if I squash your commit into my testing commit? We should be good to go after that I think.

addaleax commented:
@Fishrock123 sure, go ahead!

addaleax commented:
LGTM, by the way.

Fishrock123 commented:
Moving `ctc-agenda` to the tracking issue: https://github.com/nodejs/node/issues/6980

Fishrock123 commented:
Landing tomorrow before the release unless anyone objects. (@nodejs/ctc)

If someone could check my commit descriptions that would be appreciated. I've tried to put as much information in as possible while still having it be somewhat succinct.

addaleax commented:
More like tiny nits, but still: `ansi control characters, such at \r` contains a typo (`at`) and may be confusing in that `\r` wouldn’t be considered an *ANSI* control character… Shortening it to `when using control characters, such as \r` should work fine.

`flusing` appears to be another typo in the first commit message.

TheAlphaNerd commented:
@Fishrock123 commit message is looking good for the most part.

```
This became an issue as result of the Libuv 1.9.0 upgrade, which fixed
a different issue which in turn had side effects that bypassed a
previous "hack" which had been done to make OS X work properly.
The 1.9.0 upgrade was done in c3cec1e,
which was included in v6.0.0.
Full details of the Libuv issue that induced this are at
nodejs#6456 (comment)
```

The section above read a bit awkward to me. Maybe something like

```
This became an issue as result of the Libuv 1.9.0 upgrade. A fix to
an unrelated issue resulted in a behavior change to a hack that
in the OS X implementation which resulted in an unexpected behavior
change in node.
The 1.9.0 upgrade was done in c3cec1e,
which was included in v6.0.0.
Full details of the Libuv issue that induced this are at
nodejs#6456 (comment)
````

I think my attempt is somewhat awkward, let me know if you want to do another iteration back and forth

thefourtheye commented on test/pseudo-tty/testcfg.py:
```diff
@@ -0,0 +1,166 @@
+    f = file(self.expected)
+    # Skip initial '#' comment and spaces
+    #for line in f:
+    #  if (not line.startswith('#')) and (not line.strip()):
```
These comments can be removed

thefourtheye commented on test/pseudo-tty/testcfg.py:
```diff
@@ -0,0 +1,166 @@
+    all_tests = [current_path + [t] for t in self.Ls(self.root)]
+    result = []
+    # Skip these tests on Windows, as pseudo terminals are not available
+    # FIXME(Fishrock123): Currently this only runs properly on OS X in the
```
Is this still needed?

Fishrock123 replied:
Oh, nope. Good catch.

Fishrock123 commented:
@TheAlphaNerd Updated, mind one last pass on the commit message?

@thefourtheye I've addressed those nits, however I have not addressed some of your original python style nits. Could we do those in a separate PR to address them in both the message and these tests at the same time?

Anything else before I land this later today (us time)?

Fishrock123 commented:
CI to verify the python still works: https://ci.nodejs.org/job/node-test-pull-request/2875/

Fishrock123 commented:
added some docs notes which i forgot before.. @TheAlphaNerd

jasnell commented on test/pseudo-tty/no_dropped_stdio.js:
```diff
@@ -0,0 +1,16 @@
+// Array(num).join() "loops" one less time than may be expected.
+// \n turns into a double-byte character
+// 20 * (48 + (2))
+var out = new Array(21).join(new Array(49).join('o') + '\n');
```
```
out = `${'0'.repeat(49)}\n`.repeat(21)
```

jasnell commented on test/pseudo-tty/no_dropped_stdio.js:
```diff
@@ -0,0 +1,16 @@
+var out = new Array(21).join(new Array(49).join('o') + '\n');
+// Add the remaining 24 bytes and terminate with an 'O'.
+// This results in 1025 bytes, just enough to overflow the 1kb OS X TTY buffer.
+out += new Array(25).join('o') + 'O';
```
```
out += 'o'.repeat(25) + 'O'
```

TheAlphaNerd commented on doc/api/process.md:
```diff
@@ -1359,6 +1360,10 @@ event and that writes can block when output is redirected to a file (although
 
+Additionally, `process.stderr` and `process.stdout` are blocking when outputting
+to TTYs (terminals) on OS X as a workaround for the OS's very small, 1kb
+buffer size.
```
you may want to include that this is done to avoid output from interleaving

Fishrock123 commented:
@TheAlphaNerd, @jasnell update with nits, one last CI (hopefully): https://ci.nodejs.org/job/node-test-pull-request/2897/

TheAlphaNerd commented:
great job @Fishrock123 

This is going to help quite a few people

saghul commented:
Great work @Fishrock123, I appreaciate you took the time to go deep down the rabbit hole. Also thanks a lot to everyone who helped!

saghul commented:
I'd also like to propose "Making blocking writes great again" as the next release tagline.
